### PR TITLE
Spotaut 15748- support user management in the terraform provider

### PR DESCRIPTION
**Pull request by pripatra**
_This is the default pull request template. You can customize it by adding a `pull_request_template.md` at the root of your repo or inside the `.github` folder._

# Jira Ticket
https://spotinst.atlassian.net/browse/SPOTAUT-15748
_Include a link to your Jira Ticket_
_Example: [JIRAISS-1234](https://spotinst.atlassian.net/browse/JIRAISS-1234)_

# Demo

_Please add a recording of the feature/bug fix in work. if you added new routes, the recording should show the request and response for each new/changed route_

# Checklist:
- [ ] I have filled relevant self assessment ([NodeJS](https://docs.google.com/forms/d/e/1FAIpQLSfl14u9AOBAmxVJ272tvO7XNuXE-EMvWaGcaRZalu1UAKB7RA/viewform), [Frontend](https://docs.google.com/forms/d/e/1FAIpQLSdiBPNKH81w_EkavihVL8Uwb0j7tP8PwJmLFYm2nCOQxz-1qw/viewform), [Backend](https://docs.google.com/forms/d/e/1FAIpQLSed_PsTJ5-XIWkFL6BSDE2AQRBVPmwc3PAmHZkUn-erzVI37Q/viewform?usp=sf_link))
- [ ] I have run ESlint on my changes and fixed all warnings and errors (**NodeJS & Frontend Services**)
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have validated all the requirements in the Jira task were answered
- [ ] I have all neccessary approvals for the design/mini design of this task
- [ ] I have approved the API changes and granular permission patterns (documentation subtask) (**For public services only**)

### CODE REVIEW

File: examples/service/administration/policy/create/main.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	svc := administration.New(sess)
+
+	// Create a new context.
+	ctx := context.Background()
+
+	// Create a new group.
+	out, err := svc.CreatePolicy(ctx, &administration.CreatePolicyInput{
+		Policy: &administration.Policy{
+			Description: spotinst.String("TFPolicy"),
+			Name:        spotinst.String("Pritish"),
```
Comment (anuragsharma-123) [inline examples/service/administration/policy/create/main.go:34]: change it to "Automation-Policy"

File: examples/service/administration/policy/read/main.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	"context"
+	"github.com/spotinst/spotinst-sdk-go/service/administration"
+	"log"
+
+	"github.com/spotinst/spotinst-sdk-go/spotinst"
+	"github.com/spotinst/spotinst-sdk-go/spotinst/session"
+	"github.com/spotinst/spotinst-sdk-go/spotinst/util/stringutil"
+)
+
+func main() {
```
Comment (anuragsharma-123) [inline examples/service/administration/policy/read/main.go:13]: This is Policy Read implementation, why it is returning user.

File: examples/service/administration/policy/update/main.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	svc := administration.New(sess)
+
+	// Create a new context.
+	ctx := context.Background()
+
+	// Create a new group.
+	out, err := svc.UpdatePolicy(ctx, &administration.UpdatePolicyInput{
+		Policy: &administration.Policy{
+			PolicyID: spotinst.String("pol-c75d8c06"),
+			Name:     spotinst.String("PritishUpdated"),
```
Comment (anuragsharma-123) [inline examples/service/administration/policy/update/main.go:34]: Rename to Automation-Policy-Updated

File: examples/service/administration/user/create/main.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	// Optional spotinst.Config values can also be provided as variadic
+	// arguments to the New function. This option allows you to provide
+	// service specific configuration.
+	svc := administration.New(sess)
+
+	// Create a new context.
+	ctx := context.Background()
+
+	// Create a new group.
+	out, err := svc.CreateUser(ctx, &administration.User{
```
Comment (anuragsharma-123) [inline examples/service/administration/user/create/main.go:31]: Please use automation specific details, not personal details.

File: examples/service/administration/user/createProg/main.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+	svc := administration.New(sess)
+
+	// Create a new context.
+	ctx := context.Background()
+
+	// Create a new group.
+	out, err := svc.CreateProgUser(ctx, &administration.ProgrammaticUser{
+		Name:        spotinst.String("pritish"),
+		Description: spotinst.String("description"),
+		/*Accounts: []*administration.Account{
```
Comment (anuragsharma-123) [inline examples/service/administration/user/createProg/main.go:34]: Why Accounts is commented in this example?

File: examples/service/administration/userGroup/create/main.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+	// service specific configuration.
+	svc := administration.New(sess)
+
+	// Create a new context.
+	ctx := context.Background()
+
+	// Create a new group.
+	out, err := svc.CreateUserGroup(ctx, &administration.UserGroup{
+		Description: spotinst.String("TFUserGroup"),
+		Name:        spotinst.String("Pritish"),
```
Comment (anuragsharma-123) [inline examples/service/administration/userGroup/create/main.go:33]: Give generic names e.g. TF_Test_UserGroup

File: examples/service/administration/userGroup/create/main.go [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+	// Optional spotinst.Config values can also be provided as variadic
+	// arguments to the New function. This option allows you to provide
+	// service specific configuration.
+	svc := administration.New(sess)
+
+	// Create a new context.
+	ctx := context.Background()
+
+	// Create a new group.
+	out, err := svc.CreateUserGroup(ctx, &administration.UserGroup{
```
Comment (anuragsharma-123) [inline examples/service/administration/userGroup/create/main.go:31]: UserIDs missing in this example.

File: examples/service/administration/userGroup/update/main.go [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+	// Optional spotinst.Config values can also be provided as variadic
+	// arguments to the New function. This option allows you to provide
+	// service specific configuration.
+	svc := administration.New(sess)
+
+	// Create a new context.
+	ctx := context.Background()
+
+	// Create a new group.
+	out, err := svc.UpdateUserGroup(ctx, &administration.UserGroup{
```
Comment (anuragsharma-123) [inline examples/service/administration/userGroup/update/main.go:31]: Can we add Policy update for User GRoup, in this example.

File: service/administration/service.go [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+// Service provides the API operation methods for making requests to endpoints
+// of the Spotinst API. See this package's package overview docs for details on
+// the service.
+type Service interface {
+	ListUsers(context.Context, *ListUsersInput) (*ListUsersOutput, error)
+	CreateUser(context.Context, *User, *bool) (*CreateUserOutput, error)
+	CreateProgUser(context.Context, *ProgrammaticUser) (*CreateProgrammaticUserOutput, error)
+	ReadUser(context.Context, *ReadUserInput) (*ReadUserOutput, error)
+	ReadProgUser(context.Context, *ReadUserInput) (*ReadProgUserOutput, error)
```
Comment (anuragsharma-123) [inline service/administration/service.go:19]: Is this a separate API, I couldn't find.

Comment (pripatra) [reply service/administration/service.go:19]: the route is same but the response is slightly different so handling with a different function

File: examples/service/administration/userGroup/update/main.go (hunk 8 above)
Comment (pripatra) [reply examples/service/administration/userGroup/update/main.go:31]: not possible api is throwing error, in the api spec also only name is updated

File: examples/service/administration/policy/create/main.go (hunk 1 above)
Comment (chandra1-n) [inline examples/service/administration/policy/create/main.go:34]: change the name to generic like testUser or similar to that

File: examples/service/administration/policy/update/main.go (hunk 3 above)
Comment (chandra1-n) [inline examples/service/administration/policy/update/main.go:34]: change the name to generic

File: examples/service/administration/user/createProg/main.go [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+			{
+				Id:   spotinst.String("act-7c46c6df"),
+				Role: spotinst.String("viewer"),
+			},
+		},*/
+		Policies: []*administration.ProgPolicy{
+			{
+				PolicyId: spotinst.String("pol-c75d8c06"),
+				AccountIds: []string{
+					"act-7c46c6df",
```
Comment (chandra1-n) [inline examples/service/administration/user/createProg/main.go:44]: remove the account no. value

File: service/administration/user.go [hunk 11]
```diff
@@ -0,0 +1,466 @@
...
+	Password            *string       `json:"password,omitempty"`
+	Role                *string       `json:"role,omitempty"`
+	UserID              *string       `json:"userId,omitempty"`
+	Username            *string       `json:"username,omitempty"`
+	Type                *string       `json:"type,omitempty"`
+	Mfa                 *bool         `json:"mfa,omitempty"`
+	Policies            []*UserPolicy `json:"policies,omitempty"`
+	Tokens              []*Token      `json:"tokens,omitempty"`
+	PersonalAccessToken *string       `json:"personalAccessToken,omitempty"`
+	Id                  *int          `json:"id,omitempty"`
```
Comment (chandra1-n) [inline service/administration/user.go:30]: we have few more attributes returned in the create user response

File: service/administration/user.go [hunk 12]
```diff
@@ -0,0 +1,466 @@
...
+	nullFields      []string
+}
+
+type ListUsersInput struct{}
+
+type ListUsersOutput struct {
+	Users []*User `json:"users,omitempty"`
+}
+
+type CreateUserInput struct {
```
Comment (chandra1-n) [inline service/administration/user.go:103]: why there is no input passed

File: service/administration/userGroup.go [hunk 13]
```diff
@@ -0,0 +1,293 @@
...
+	nullFields      []string
+}
+
+type ListUserGroupsInput struct{}
+
+type ListUserGroupsOutput struct {
+	UserGroups []*UserGroup `json:"userGroups,omitempty"`
+}
+
+type CreateUserGroupInput struct {
```
Comment (chandra1-n) [inline service/administration/userGroup.go:53]: why there is no input passed

File: examples/service/administration/user/createProg/main.go (hunk 10 above)
Comment (pripatra) [reply examples/service/administration/user/createProg/main.go:44]: added a dummy value

File: service/administration/userGroup.go [hunk 14]
```diff
@@ -0,0 +1,293 @@
...
+	"github.com/spotinst/spotinst-sdk-go/spotinst/util/jsonutil"
+	"github.com/spotinst/spotinst-sdk-go/spotinst/util/uritemplates"
+)
+
+type UserGroup struct {
+	Description *string            `json:"description,omitempty"`
+	Name        *string            `json:"name,omitempty"`
+	Policies    []*UserGroupPolicy `json:"policies,omitempty"`
+	UserIds     []string           `json:"userIds,omitempty"`
+	UserGroupId *string            `json:"id,omitempty"`
```
Comment (chandra1-n) [inline service/administration/userGroup.go:20]: more attributes under get user group details to be added here

File: examples/service/administration/userGroup/create/main.go [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+		Description: spotinst.String("TFUserGroup"),
+		Name:        spotinst.String("Pritish"),
+		Policies: []*administration.UserGroupPolicy{
+			{
+				AccountIds: []string{
+					"act-7c46c6df",
```
[the remark this answers is not in the document]
Comment (pripatra) [reply examples/service/administration/userGroup/create/main.go:37]: added dummy value

File: service/administration/user.go (hunk 11 above)
Comment (pripatra) [reply service/administration/user.go:30]: added

File: service/administration/user.go [hunk 16]
```diff
@@ -0,0 +1,466 @@
...
+	TokenId         *int       `json:"tokenId,omitempty"`
+	TokenLastDigits *string    `json:"tokenLastDigits,omitempty"`
+
+	forceSendFields []string
+	nullFields      []string
+}
+
+type ListUsersInput struct{}
+
+type ListUsersOutput struct {
```
Comment (chandra1-n) [inline service/administration/user.go:99]: get users has groupNames which is not covered in user struct

File: service/administration/user.go [hunk 17]
```diff
@@ -0,0 +1,466 @@
...
+
+type CreateProgrammaticUserOutput struct {
+	ProgrammaticUser *ProgrammaticUser `json:"user,omitempty"`
+}
+
+type ReadUserInput struct {
+	UserID *string `json:"userId,omitempty"`
+}
+
+type ReadUserOutput struct {
```
Comment (chandra1-n) [inline service/administration/user.go:121]: get user details has groups object in the response. which is not covered in user struct

File: service/administration/user.go [hunk 18]
```diff
@@ -0,0 +1,466 @@
...
+	// empty value appearing in NullFields will be sent to the server as
+	// null. It is an error if a field in this list has a non-empty value.
+	// This may be used to include null fields in Patch requests.
+	nullFields []string
+}
+
+type UserPolicy struct {
+	PolicyId   *string `json:"policyId,omitempty"`
+	PolicyName *string `json:"policyName,omitempty"`
+	PolicyType *string `json:"policyType,omitempty"`
```
Comment (chandra1-n) [inline service/administration/user.go:52]: if this is part of get user details response then accountIds missing here

File: service/administration/user.go [hunk 19]
```diff
@@ -0,0 +1,466 @@
...
+	Role *string `json:"role,omitempty"`
+
+	forceSendFields []string
+	nullFields      []string
+}
+
+type Token struct {
+	Name            *string    `json:"name,omitempty"`
+	CreatedAt       *time.Time `json:"createdAt,omitempty"`
+	PolicyType      *string    `json:"policyType,omitempty"`
```
Comment (chandra1-n) [inline service/administration/user.go:89]: couldn't find policyType as part of this Token object, please re-check

**Review by chandra1-n — CHANGES_REQUESTED** [review]
please fix the review comments

File: service/administration/user.go [hunk 20]
```diff
@@ -0,0 +1,471 @@
...
+	forceSendFields []string
+	nullFields      []string
+}
+
+type ListUsersInput struct{}
+
+type ListUsersOutput struct {
+	Users []*User `json:"users,omitempty"`
+}
+
```
Comment (chandra1-n) [inline service/administration/user.go:113]: CreateUserInput and CreateProgrammaticUserInput missing

File: service/administration/userGroup.go [hunk 21]
```diff
@@ -0,0 +1,305 @@
...
+	AccountIds []string `json:"accountIds,omitempty"`
+	PolicyId   *string  `json:"policyId,omitempty"`
+	PolicyName *string  `json:"policyName,omitempty"`
+	PolicyType *string  `json:"policyType,omitempty"`
+
+	forceSendFields []string
+	nullFields      []string
+}
+
+type UserGroupUser struct {
```
Comment (chandra1-n) [inline service/administration/userGroup.go:53]: change the struct name to getGroupUsers

File: service/administration/userGroup.go [hunk 22]
```diff
@@ -0,0 +1,305 @@
...
+	// requests with the JSON null value. By default, fields with empty
+	// values are omitted from API requests. However, any field with an
+	// empty value appearing in NullFields will be sent to the server as
+	// null. It is an error if a field in this list has a non-empty value.
+	// This may be used to include null fields in Patch requests.
+	nullFields []string
+}
+
+type UserGroupPolicy struct {
+	AccountIds []string `json:"accountIds,omitempty"`
```
Comment (chandra1-n) [inline service/administration/userGroup.go:44]: get user group details has this as account names, check in postman once

File: service/administration/userGroup.go [hunk 23]
```diff
@@ -0,0 +1,305 @@
...
+type ReadUserGroupOutput struct {
+	UserGroup *UserGroup `json:"userGroup,omitempty"`
+}
+
+type UpdateUserGroupInput struct {
+	UserGroupID *string `json:"id,omitempty"`
+}
+
+type UpdateUserGroupOutput struct {
+	UserGroup *UserGroup `json:"userGroup,omitempty"`
```
Comment (chandra1-n) [inline service/administration/userGroup.go:85]: update user group details doesn't return anything, check once if it's really required

**Review by chandra1-n — CHANGES_REQUESTED** [review]
please fix review comments

File: service/administration/user.go (hunk 20 above)
Comment (pripatra) [reply service/administration/user.go:113]: Not needed we are directly using User and ProgramatticUser structs for the same

File: service/administration/userGroup.go (hunk 22 above)
Comment (pripatra) [reply service/administration/userGroup.go:44]: It is incorrect in API spec, the response is retrurning account Ids not names

File: service/administration/userGroup.go (hunk 23 above)
Comment (pripatra) [reply service/administration/userGroup.go]: removed it and changed update user grou pimplementation

**Review by chandra1-n — APPROVED** [review]

**Pull request merged**